### PR TITLE
Fix CanMatchPreFilterSearchPhaseTests#testSortShards

### LINKED ISSUE #70933: [CI] CanMatchPreFilterSearchPhaseTests testSortShards failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/e3co62woepqba/tests/:server:test/org.elasticsearch.action.search.CanMatchPreFilterSearchPhaseTests/testSortShards

**Reproduction line:**
`./gradlew ':server:test' --tests "org.elasticsearch.action.search.CanMatchPreFilterSearchPhaseTests.testSortShards" -Dtests.seed=BA740A4C6B31F62D -Dtests.security.manager=true -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dtests.locale=sq-AL -Dtests.timezone=Europe/Oslo -Druntime.java=11`

**Applicable branches:**
master

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.action.search.CanMatchPreFilterSearchPhaseTests&tests.test=testSortShards

**Failure excerpt:**
```
  java.lang.AssertionError: expected:<true> but was:<false>

  at __randomizedtesting.SeedInfo.seed([BA740A4C6B31F62D:DCE66A962CB8FCD7]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.failNotEquals(Assert.java:834)
  at org.junit.Assert.assertEquals(Assert.java:118)
  at org.junit.Assert.assertEquals(Assert.java:144)
  at org.elasticsearch.action.search.CanMatchPreFilterSearchPhaseTests.testSortShards(CanMatchPreFilterSearchPhaseTests.java:362)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:566)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:834)

```

### PR DESCRIPTION
**Pull request by jimczi**
Fix test to handle the case where all shards are skipped.
We ensure that we don't skip all shards in the can match phase
in order to be able to produce an empty result for aggs.

Closes #70933

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhaseTests.java [hunk 1]
```diff
@@ -356,7 +356,10 @@ public void run() {
                 .sorted(Comparator.comparing(minAndMaxes::get, MinAndMax.getComparator(order)).thenComparing(shardIds::get))
                 .map(shardIds::get)
                 .toArray(ShardId[]::new);
-
+            if (shardToSkip.size() == expected.length) {
+                // we need at least one shard to produce the empty result for aggs
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhaseTests.java:360]: For my own learning: this is the special case mentioned in CanMatchPreFilterSearchPhase#getIterator where we add at least one shard index to the set to be queried? And that is the reason why we need to remove that one from the skipped set here in the test?

**Comment by cbuescher** [comment]
I left one question for my own learning but think I understand, just for confirmation. LGTM.

**Review by cbuescher — APPROVED** [review]

**Review by dnhatn — APPROVED** [review]
LGTM2

File: server/src/test/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhaseTests.java (hunk 1 above)
Comment (jimczi) [reply server/src/test/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhaseTests.java:360]: yep, you got it right.

**Pull request merged**